File: src/game/map/Layer.java
package game.map;

import game.graphics.SpriteBatch;
import game.tiles.Tile;
import game.tiles.Tiles;
import game.world.World;

public class Layer {

    private int mapWidth;
    private int mapHeight;

    private Tile[] tiles;
    private int[] durabilty;
    private int[] direction;

    private SpriteBatch spriteBatch;


    //TODO because we are going to procedurally generate the maps we need some kind of map loader/creator
    public Layer(SpriteBatch spriteBatch, int width, int height) {
        this.spriteBatch = spriteBatch;
        mapWidth = width;
        mapHeight = height;
        tiles = new Tile[width * height];
        durabilty = new int[width * height];
        direction = new int[width * height];

        for (int i = 0; i < tiles.length; i++) {
            tiles[i] = Tiles.treeTile;
            durabilty[i] = 100;
        }
    }

    //TODO We might want to add more options that just render at x/y like zoom, transparency...
    public void render(int offsetX, int offsetY) {
        int endX = mapWidth;
        int endY = mapHeight;

        spriteBatch.begin();
        spriteBatch.setOffsets(World.getPlayerXOffset(), World.getPlayerYOffset());
        for (int y = offsetY; y < endY; y++) {
            for (int x = offsetX; x < endX; x++) {
                tiles[getPosition(x, y)].render(tileToPixel(x), tileToPixel(y), spriteBatch);
            }
        }
        spriteBatch.end();
    }

    public void update(int xPosition, int yPosition, long delta) {
        tiles[getPosition(xPosition, yPosition)].update(delta);
    }

    public void remove(int xPosition, int yPosition) {

    }

    public void interact(int xPosition, int yPosition) {
        if (inBounds(xPosition, yPosition)) {
            int position = getPosition(xPosition, yPosition);

            durabilty[position] -= 20;

            if (durabilty[position] <= 0)
                tiles[position] = Tiles.greenTile;
        }
    }

    private boolean inBounds(int x, int y) {
        return x >= 0 && y >= 0 && x < mapWidth && y < mapHeight;
    }

    private int getPosition(int x, int y) {
        return x + y * mapWidth;
    }

    public static int tileToPixel(int tile) {
        return tile << 6;
    }

    public static float pixelToTile(int pixel) {
        return pixel >> 6;
    }
}

File: src/game/map/Layer.kt
package game.map

import game.entity.Player
import game.graphics.SpriteBatch
import game.math.TilePosition
import game.math.Vector2
import game.tiles.Tile
import game.tiles.Tiles
import game.util.Util.pixelToTile
import game.util.Util.tileToPixel
import game.world.World
import org.lwjgl.opengl.Display

//TODO because we are going to procedurally generate the maps we need some kind of map loader/creator
class Layer(private val world: World, private val spriteBatch: SpriteBatch, private val mapWidth: Int, private val mapHeight: Int) {

    private val tiles: Array<Tile?>
    private val durability: IntArray
    private val direction: IntArray


    init {
        tiles = arrayOfNulls<Tile>(mapWidth * mapHeight)
        durability = IntArray(mapWidth * mapHeight)
        direction = IntArray(mapWidth * mapHeight)

        for (i in 0..tiles.size - 1) {
            tiles[i] = Tiles.treeTile
            durability[i] = 100
        }
    }

    //TODO We might want to add more options that just render at x/y like zoom
    fun render(offsetX: Int, offsetY: Int) {
        var endX = (pixelToTile(Display.getWidth()) - pixelToTile(World.playerXOffset) + 1).toInt()
        var endY = (pixelToTile(Display.getHeight()) - pixelToTile(World.playerYOffset) + 1).toInt()

        if (offsetX > 0)
            endX += offsetX
        if (offsetY > 0)
            endY += offsetY

        spriteBatch.begin()
        spriteBatch.setOffsets(World.playerXOffset, World.playerYOffset)
        for (y in offsetY..endY) {
            for (x in offsetX..endX - 1) {
                if (inBounds(x, y))
                    tiles[getTileIndex(x, y)]?.render(tileToPixel(x).toInt(), tileToPixel(y).toInt(), spriteBatch)
            }
        }
        spriteBatch.end()
    }

    fun harvest(player: Player, position: TilePosition) {
        if (inBounds(position)) {
            val index = position.index(mapWidth)

            durability[index] -= 20

            if (durability[index] <= 0) {
                tiles[index] = Tiles.greenTile
                tiles[index]?.brake(world, Vector2(position.pixelX(), position.pixelY()))
            }
        }
    }

    private fun inBounds(x: Int, y: Int): Boolean {
        return x >= 0 && y >= 0 && x < mapWidth && y < mapHeight
    }

    private fun inBounds(tilePosition: TilePosition): Boolean {
        return tilePosition.x >= 0 && tilePosition.y >= 0 && tilePosition.x < mapWidth && tilePosition.y < mapHeight
    }

    private fun getTileIndex(x: Int, y: Int): Int {
        return x + y * mapWidth
    }

    private fun getTileIndex(tilePosition: TilePosition): Int {
        return (tilePosition.x + tilePosition.y * mapWidth).toInt()
    }

    companion object {
    }
}
